graphix/sim/density_matrix.py

```diff
@@ -172,20 +179,19 @@ def expectation_single(self, op, i):
         if op.shape != (2, 2):
             raise ValueError("op must be 2x2 matrix.")
 
-        st1 = deepcopy(self)
```

Comment:
I feel it's more idiomatic to use `deepcopy`.

Reply:
`normalize` already copy the contents of the matrix, so it's preferable just to do a shallow copy here, instead of copying the whole matrix twice (first with `deepcopy`, and just after with `normalize`).

Reply:
I see, but then we should use `copy.copy` instead: I cannot see an obvious reason why `.copy` should be provided as a dedicated method.

Reply:
It’s quite common in Python for objects to provide their own copy methods—examples include `set.copy()`, `list.copy()`, and `dict.copy()`. These methods handles the copying process in a way that is specific to the object’s internal structure. Relying on the copy module (`copy.copy()` or `copy.deepcopy()`) can sometimes expose or depend on internal details of the object that we might not want to reveal.

For instance, in the case of `NodeIndex`, the `copy` method is designed to copy one level deeper than a standard shallow copy, ensuring that internal collections are duplicated into new instances rather than being shared between the original and the copy. In the case of back-ends, they currently only contain their state object, but that's an implementation detail that the caller doesn't have to know about.

Reply:
> `set.copy()`, `list.copy()`, and `dict.copy()`

To my understanding, these methods serves as the "leaves" of object trees, which we are closely tied to the C implementations: they are special and inevitable.
For our implementation, they're unnecessary as we're coding in the Python layer.

> designed to copy one level deeper than a standard shallow copy

I feel it isn't conventional or idiomatic.
It might be no problem if deepcopy were to be a free lunch, but sadly not.
We may need to discuss whether slight conciseness deserves unconventionality and computational cost.

Reply:
@thierry-martinez does `normalize` really do deepcopy?  even if it does, I would be up for `copy.copy` here to be clear to anyone that it's a shallow copy. I'm not against `.copy()` method but I personally many not use it to be sure what I am doing where.

Reply:
Fixed in e9f2dce. I removed the copy methods from `Statevec`, `DensityMatrix`, and `NodeIndex`, and explicitly used `copy.copy`. I believe that choosing between `copy` and `deepcopy` assumes knowledge of the internals of the classes being copied, but I understand the preference for being explicit.

As for `normalize`, it indeed normalizes the array, but it's similar to a `deepcopy` in the following ways:

    There is no aliasing between the normalized array and the original array.
    The process iterates over each element of the array.

Importantly, there is no need to `deepcopy` before `normalize` (or `evolve`). If `d1` is a shallow copy of `d0` (meaning they share a reference to the same numpy array, i.e., `d0.rho is d1.rho`), then `d1.normalize()` changes `d1.rho` to reference the newly normalized array, while `d0.rho` remains unmodified.

Reply:
Looks good. Thank you!